[BE] feat: 홈화면 카테고리별 히어릿 조회 API 구현 #212

Byesol opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->
- 홈화면에 표시할 3개의 카테고리의 최신 Hearit 게시글 5개씩을 반환하는 API를 구현.
- 가장 오래된 3개 카테고리를 기준으로 `Category`와 연관된 최신 Hearit 목록을 제공합니다.
> UI 변경이 있을 경우 첨부

## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->
- `HearitControllerTest.readHomeCategoriesHearit()`
- `HearitServiceTest.getHomeCategoryHearits()`
- `HearitRepositoryTest.findTop5ByCategory()`
## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#212 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 최신 3개 카테고리별로 최대 5개의 히어릿을 그룹화하여 반환하는 새로운 API 엔드포인트가 추가되었습니다.
  * 카테고리별로 묶인 히어릿 목록을 조회할 수 있습니다.

* **버그 수정**
  * 리소스 미발견 예외 처리 시 불필요한 에러 로그가 제거되었습니다.

* **테스트**
  * 카테고리별 히어릿 그룹화, 랜덤 조회, 카테고리별 조회 등 신규 기능에 대한 테스트가 추가되었습니다.
  * 카테고리 조회 기능에 대한 저장소 테스트가 추가되었습니다.

* **환경 설정**
  * Jackson의 기본 타임존이 Asia/Seoul로 설정되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Byesol committed:
[BE] feat: 카테고리 3개 조회 메서드 구현 #212

Byesol committed:
[BE] feat: 특정 카테테고리의 히어릿 5개 조회 구현 #212

Byesol committed:
[BE] feat: 홈 화면 카테고리 및 히어릿 조회 기능 구현 #212

- 3개의 카테고리 조회
- 현재 카테고리 조회 기준은 오래된 순
- 카테고리별 5개의 히어릿 조회 (최신순 5개)

Byesol committed:
[BE] test : 홈 화면 카테고리 히어릿 조회 기능 테스트 구현 #212

- 인수 테스트 구현
- 서비스 테스트 구현

Byesol committed:
[BE] Merge remote-tracking branch 'origin/develop-be' into feat/#212-home-category

# Conflicts:
#	backend/src/main/java/com/onair/hearit/application/HearitService.java
#	backend/src/main/java/com/onair/hearit/infrastructure/HearitRepository.java
#	backend/src/test/java/com/onair/hearit/application/HearitServiceTest.java

Byesol committed:
[BE] refactor: 병합하며 생긴 필요없는 소스코드 제거 #212

Byesol committed:
[BE] test: 카테고리별 원하는 히어릿 개수만큼 조회 테스트 구현 #212

Byesol committed:
[BE] refactor: 필요없어진 상수 제거 #212

Byesol committed:
[BE] refactor: 메서드명 findByCategory로 변경 #212

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -66,4 +71,15 @@ public List<RecommendHearitResponse> getRecommendedHearits() {
+
+    public List<HomeCategoryHearitResponse> getHomeCategoryHearits() {
+        Pageable pageable = PageRequest.of(0, CATEGORY_HEARIT_COUNT);
+        List<Category> categories = categoryRepository.findTop3ByOrderByIdAsc();
```
Top3라고 하지 말고 서비스에서 파라미터를 넘기는 건 어떤가요

Byesol replied:
limit를 이용하여 파라미터 넘기는 방식으로 변경했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/7f4cbdc8bf68c977c3880f0283391a527862d07e

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -66,4 +71,15 @@ public List<RecommendHearitResponse> getRecommendedHearits() {
     }
+
+    public List<HomeCategoryHearitResponse> getHomeCategoryHearits() {
+        Pageable pageable = PageRequest.of(0, CATEGORY_HEARIT_COUNT);
```
페이징이 필요한가요? 페이징 기능이 모두 필요한 게 아니라 5개까지만 가져오는 기능을 위한 거라면 단순 limit만 써도 충분하지 않을까요? 물론 페이징을 쓰면 JpaRepository에서 간편하게 구현해준다는 것은 알지만~

Byesol replied:
기존의 작성되어있던
```
    public List<RecommendHearitResponse> getRecommendedHearits() {
        Pageable pageable = PageRequest.of(0, RECOMMEND_HEARIT_COUNT);
        return hearitRepository.findRandom(pageable).stream()
                .map(RecommendHearitResponse::from)
                .toList();
    }
```

코드와 동작 방식이 유사하다고 생각해서 형식을 그대로 맞췄습니다.  전 두 방식의 큰 차이는 없어 컨벤션을 맞추면 된다고 생각하는데 만약 반영해 수정한다면  기존에 작성되어있던 'getRecommendedHearits' 도 수정할까요?

yuyoungrhee replied:
이 부분은 팀원들이랑 좀 더 논의해봐야될 것 같네요. 우선 저는 페이징을 쓸거면 페이징답게 써야된다는 입장입니다!!

JO-eusan replied:
Recommend 메서드에서 페이징을 사용했던 이유는, 단순 메서드가 간단해져서 사용했었습니다. 특히, 동적으로 값을 파라미터로 넣어서 쿼리를 생성했으면 좋겠다는 의견이 있었어서 대표적인 Pageable을 사용했습니다.
만약 동적으로 값을 받지 않아도 된다면 `First`, `Top`으로 생성하는 JPA query 메서드를 사용하거나 아니면 limit 키워드를 사용해야 할 것 같습니다.
(JPQL의 limit 키워드는 hibernate 버전에 따라 사용 여부 체크는 해봐야될 것 같습니다.)

Byesol replied:
논의 후 페이징 request로 값을 받는 게 아닌 jpql에서 파라미터로 limit 값을 받는 방식으로변경했습니 다.
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/7f4cbdc8bf68c977c3880f0283391a527862d07e

yuyoungrhee commented on backend/src/test/java/com/onair/hearit/infrastructure/HearitRepositoryTest.java:
```diff
@@ -130,6 +131,34 @@ void searchByTerm_avoidDuplicateWhenTitleAndKeywordMatch() {
+        // given
+        Category category1 = dbHelper.insertCategory(TestFixture.createFixedCategory());
+        Category category2 = dbHelper.insertCategory(TestFixture.createFixedCategory());
+
```
여기도 아래처럼

//category1에 6개 저장 이렇게 적어주면 좋을 거 같아용

Byesol replied:
추가했습니다!

https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/9184a09755adbe0eb16bdcf6bbe9a407b26db6d3

yuyoungrhee reviewed (commented):
빠르게 구현해주셨네요 수고하셨습니다! 몇가지만 확인해주세용

gabean13 commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -66,4 +71,15 @@ public List<RecommendHearitResponse> getRecommendedHearits() {
     }
+
+    public List<HomeCategoryHearitResponse> getHomeCategoryHearits() {
+        Pageable pageable = PageRequest.of(0, CATEGORY_HEARIT_COUNT);
```
HomeCateogryHearit도 있고 CategoryHearit도 있어가지고 좀 헷갈리는 것 같아요ㅠㅠ 사실 저도 베스트 네이밍은 잘 모르겠는데 카테고리에 따라서 히어릿들을 한 섹션으로 묶어서 준다는 의미가 좀 더 드러나면 좋을 것 같아요

그리고 PagedResponse로 넘겨주는거 어떤가요?(사실 이부분은 안드로이드랑 얘기해서 현재 리스트의 끝까지 갔을 때 더 받아올건지, 그냥 끝인지 정해야할것같아용)

Byesol replied:
1. 팀 논의 후 GroupedHearitsWithCategoryResponse 와 HearitResponse로 변경했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/8d0b708730f654653fae8de6929621f476058551
2. PagedResponse는 논의 후 해당 api에서는 pagedResponse가 아닌 List로 받기로 했습니다!

gabean13 commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -66,4 +71,15 @@ public List<RecommendHearitResponse> getRecommendedHearits() {
+                    return HomeCategoryHearitResponse.from(category, hearits);
+                })
+                .toList();
+    }
```
이 부분 가독성이 조금 떨어지는 것 같아서 하나의 Category에 대해 Hearit List를 가져오는 메서드로 분리하는거 어떤가용?!!

그리고 지금은 고정된 Category지만 나중에는 사용자에 맞게 추천할거니까 todo라도 써주면 좀 더 이해하기 편할것같아요. 임시로 만들어놓은거니까 써놓으면 나중에 수정하기 편할 것 같아요

Byesol replied:
메서드 분리 및 todo 작성 완료했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/7f4cbdc8bf68c977c3880f0283391a527862d07e
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/c86b46c4e3415b77574b6b288db659f9742d3cd7

gabean13 reviewed (commented):
고생하셨어용

JO-eusan commented on backend/src/main/java/com/onair/hearit/dto/response/HomeCategoryHearitResponse.java:
```diff
@@ -0,0 +1,47 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public record HomeCategoryHearitResponse(
```
어떻게 보면 `HomeCategoryHearitResponse`와 `CategoryHearitResponse`가 같이 있는데 한 DTO 클래스의 책임이 커보입니다. 

특히 UI적으로 생각해보면, 꼭 동시에 한 api에서 정보를 줘야할까요?
저는 개인적으로
1. 추천 카테고리에 대한 정보 (기존 CategoryResponse)를 List 형태로 주고
2. 해당 카테고리 id를 다시 받아서 해당 카테고리의 히어릿을 요청하는 흐름이 가독성이 좋고 이해가 더 자연스럽다고 생각합니다.
이렇게 되면 하나의 api에서는 `List<CategoryResponse>`를, 다음 api에서는 `List<HearitResponse>`로 반환값이 명확해질 것 같아요!
벡터의 의견도 궁금합니다.

Byesol replied:
1-1. 일단 저는 HomeCategoryHearitResponse 는 단일 책임 원칙을 위반하지 않는다고 생각해요.
HomeCategoryHearitResponse는 홈 화면의 카테고리별 히어릿이라는 하나의 응답 목적을 가지고  CategoryHearitResponse는 여기서만 사용하는 이를 구성하는 하위 구조일 뿐이기 때문에 응집도를 해치지 않는다 생각해요,
1-2. 또한, 내부의 클래스를 분리하는 것도 고려해봤지만  아래의 pr에서 언급된 것처럼 [HearitDetailResponse.java](https://github.com/woowacourse-teams/2025-hEARit/pull/217/files/f5613e5a6d68f93c2ac8445416f3c08f7ba3753d#diff-df4cfa1a017c361d33047e81a49acdda1fdb8c3b55f5ab6d631a4cc2d3ae45dd)
하나의 api에서 하나의 dto만 사용하는 저희 구조에서는 굳이 내부에서만 사용할 dto를 나눌 필요없다는 가콩의 말에 동의해 저 역시 컨벤션을 맞춰 내부 클래스를 분리하지 않았습니다.

2.  api 분리 관련해서는  이전에 안드로이드 측과 api 관련해 협의해봤을 때, 각 카테고리마다 ID를 받아서 다 요청하는 방식이  비효율적이기에 한 번의 api 로 카테고리명, 카테고리 id, 카테고리별 히어릿 리스트를 다 받고 싶다는 의견을 들었기에 결론적으로 현재의 형태와 같이 api를 작성했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/infrastructure/HearitRepository.java:
```diff
@@ -29,4 +30,7 @@ OR LOWER(k.name) LIKE LOWER(CONCAT('%', :searchTerm, '%'))
             """, nativeQuery = true)
     Page<Hearit> searchByTerm(@Param("searchTerm") String searchTerm, Pageable pageable);
+
+    @Query("SELECT h FROM Hearit h WHERE h.category.id = :categoryId ORDER BY h.createdAt DESC")
```
이 쿼리는 한 눈에 잘 안들어와서 줄바꿈이 필요해 보여요!

Byesol replied:
줄바꿈 완료했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/9184a09755adbe0eb16bdcf6bbe9a407b26db6d3

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -66,4 +71,15 @@ public List<RecommendHearitResponse> getRecommendedHearits() {
+        List<Category> categories = categoryRepository.findTop3ByOrderByIdAsc();
+        return categories.stream()
+                .map(category -> {
+                    List<Hearit> hearits = hearitRepository.findByCategory(category.getId(), pageable);
```
아래 가콩이 이야기해줬지만, 임시이고 추후 추천 알고리즘 개선에서 반영 될 거라면 주석을 달아두면 좋을 것 같아요!
특히, 이 부분에선 Pageable이 불필요하다고 느껴지긴 하네요.

Byesol replied:
주석 추가완료했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/presentation/HearitController.java:
```diff
@@ -76,4 +77,11 @@ public ResponseEntity<PagedResponse<HearitSearchResponse>> searchHearitsByTitle(
         return ResponseEntity.ok(response);
     }
+
+    @Operation(summary = "홈화면용 고정 카테고리별 히어릿 조회", description = "고정된 3개 카테고리별로 최신 히어릿 5개를 반환합니다.")
```
View에 치우치는 `summary`같다는 느낌이 있어서 '특정 카테고리 별 히어릿 조회'라는 말은 어떠신가요?
URI도 마찬가지로 `home`이라는 단어가 안들어가는게 좋을 것 같습니다.
벡터는 어떻게 생각하시나요?

yuyoungrhee replied:
`getGroupedHearitsByCategory()` Let's go

Byesol replied:
홈화면을 모두 제거하고 팀 논의 후 결정된  getGroupedHearitsByCategory를 사용하고 '카테고리별로 그룹화된 히어릿들 조회;로 변경했습니다
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/8460245a7f7f0c39756faf310f969730051df5e2

JO-eusan commented on backend/src/test/java/com/onair/hearit/application/HearitServiceTest.java:
```diff
@@ -133,4 +138,33 @@ void getRecommendedHearits() {
+    @Test
+    @DisplayName("홈 카테고리의 히어릿들을 키워드와 함께 조회할 수 있다.")
+    void getHomeCategoryHearits() {
+        // given
```
이 부분은 카테고리가 바뀔 때마다 개행을 넣는건 어떨까요? 
컨벤션으로 정해진건 아니지만...

Byesol replied:
개행 추가했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/9184a09755adbe0eb16bdcf6bbe9a407b26db6d3

JO-eusan commented on backend/src/test/java/com/onair/hearit/presentation/HearitControllerTest.java:
```diff
@@ -197,6 +198,41 @@ void searchHearitsWithInvalidParams() {
     }
 
+    @Test
+    @DisplayName("홈 카테고리 히어릿 조회 시, 3개의 카테고리와 히어릿을 반환한다.")
```
'홈 카테고리'라는 용어보다는 '특정 카테고리'가 좋을 것 같습니다.

Byesol replied:
홈 카테고리라는 이름이 적절하지 못하다는 의견에 공감합니다.
'특정 카테고리'는 기존에 존재하던 기능과 헷갈릴 수 있어 '카테고리별로 그룹화된 히어릿들 조회'라는 이름으로 변경했습니다!

JO-eusan reviewed (changes requested):
벡터 코드 확인했습니다!
api 명명 등에 대해서 리뷰 남겼으니 확인 부탁드려요!

Byesol committed:
[BE] refactor: 리소스를 찾을 수 없을 때 로깅 제거 #212

Byesol committed:
[BE] chore: 서버 시간대 설정(아시아/서울) #212

Byesol committed:
[BE] refactor: pageable request 가 아닌 limit를 통해 개수 지정하게 변경 #212

-  getHomeCategoryHearits 메서드 분리

Byesol committed:
[BE] refactor: HEARITS_PER_CATEGORY 상수 이름 변경 #212

Byesol committed:
[BE] refactor: hearitCategoryResponse의 클래스명 변경 #212

- GroupedHearitsWithCategoryResponse 로 변경

Byesol committed:
[BE] refactor: hearitCategoryResponse의 클래스명 변경 #212

- GroupedHearitsWithCategoryResponse 로 변경
- 내부 클래스 명 CategoryHearitResponse에서 HearitResponse 로 변경
- 내부 클래스에서 중복되고 불필요한 카테고리 정보 제거

Byesol committed:
[BE] chore: 사용자에 맞는 카테고리 추천에 TODO 작성 #212

Byesol committed:
[BE] refactor: 메서드명 getGroupedHearitsByCategory로 변경 #212

Byesol committed:
[BE] refactor: 테스트 개행 처리 및 한글 테스트명 변경 #212

Byesol committed:
[BE] refactor: 카테고리별 그룹화된 히어릿들 조회 url 경로 변경 #212

Byesol committed:
[BE] refactor: 가독성을 위한 주석 추가, 줄바꿈 처리 #212

- 쿼리 줄바꿈 처리
- 테스트에 주석 추가

Byesol committed:
[BE] test: categoryRepository의 findOldest 테스트 추가 #212

Byesol committed:
[BE] test: hearitRepository 의 findRandom 테스트 추가 #212

Byesol committed:
[BE] Merge remote-tracking branch 'origin/develop-be' into feat/#212-home-category

# Conflicts:
#	backend/src/main/resources/application-dev.properties

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/dto/response/GroupedHearitsWithCategoryResponse.java:
```diff
@@ -0,0 +1,43 @@
+                .toList();
+    }
+
+    public record HearitResponse(
```
inner record는 private으로 하면 좋습니다!

Byesol replied:
private처리 완료했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/da8a73fbef38fb9a19f63d4e2114f5e4d9dc64be

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/dto/response/GroupedHearitsWithCategoryResponse.java:
```diff
@@ -0,0 +1,43 @@
+        );
+    }
+
+    private static List<HearitResponse> getHearits(final List<Hearit> hearits) {
```
해당 메서드는 get의 역할이 아니라 mapping하고 있는 것 같은데요, `mapToResponses` 같은 네이밍이 더 좋을 것 같습니다. 아니면 `toHearitResponses` ?

Byesol replied:
메서드 역할에 맞게 이름 변경했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/da8a73fbef38fb9a19f63d4e2114f5e4d9dc64be

yuyoungrhee reviewed (changes requested):
잘 반영된 것 같습니다! 두 개만 확인해주세용

Byesol committed:
[BE] refactor: inner class private 처리, dto 내메서드명 변경  #212

- inner record는 private으로 변경
- getHearits가 아닌 mapToHearitResponses로 메서드명 올바르게 변경

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -25,10 +28,13 @@
 
     private static final int RECOMMEND_HEARIT_COUNT = 5;
+    private static final int CATEGORY_COUNT = 3;
+    private static final int HEARITS_PER_CATEGORY = 5;
```
`CATEGORY_COUNT`, `HEARITS_PER_CATEGORY`라는 상수 이름이 모호한 것 같습니다.
홈 화면에서 대표적으로 보일 카테고리와 컨텐츠니까 `MAIN_CATEGORY_COUNT`나 `PRIMARY_...` 느낌은 어떨까요?

Byesol replied:
상수 이름에 GROUPED를 붙여 모호성을 줄였습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/d6817dcc2b0367c4c2baf6c3d069dd4d8bc6b113

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -61,9 +67,21 @@ private RandomHearitResponse toRandomHearitResponse(Hearit hearit, Long memberId
                 .toList();
     }
+
+    //TODO 사용자에 맞는 카테고리 추천
```
함수 자체가 변하는 것이 아니라 함수의 로직이 변하는 방향이므로, 주석을 함수 내부에 작성하면 의도를 알아보기 좋을 것 같습니다.

Byesol replied:
주석위치 내부로 이동했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/d6817dcc2b0367c4c2baf6c3d069dd4d8bc6b113

JO-eusan commented on backend/src/main/java/com/onair/hearit/infrastructure/CategoryRepository.java:
```diff
@@ -1,7 +1,13 @@
 
 public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    @Query("SELECT c FROM Category c ORDER BY c.id ASC LIMIT :limit")
```
`limit`는 동사이므로 `maxSize`나 `count`가 좋을 것 같습니다.

Byesol replied:
limit는 명사로 쓰일 수 있긴 하지만 'LIMIT :limit'라는 구문이 혼동을 줄 수 있다는 의견에 공감해 size 로 변경했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/e3187ac4b6f2dde619553740538b506468473abd
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/16cf6fa36d5fc830d3b4ec8d2321cda53b84ddc5

JO-eusan commented on backend/src/main/java/com/onair/hearit/infrastructure/HearitRepository.java:
```diff
@@ -16,6 +17,9 @@ public interface HearitRepository extends JpaRepository<Hearit, Long> {
     Page<Hearit> findRandom(Pageable pageable);
 
+    @Query("SELECT h FROM Hearit h ORDER BY function('RAND') LIMIT :limit")
+    List<Hearit> findRandom(@Param("limit") int limit);
```
여기도 파라미터 명이 동사입니다.

Byesol replied:
size로 변경완료!!!!!!!!!
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/e3187ac4b6f2dde619553740538b506468473abd
https://github.com/woowacourse-teams/2025-hEARit/pull/218/commits/16cf6fa36d5fc830d3b4ec8d2321cda53b84ddc5

JO-eusan reviewed (changes requested):
벡터~ 간단한 컨벤션 정도 리뷰 남겼습니다. 확인 부탁드려요

Byesol committed:
[BE] refactor: 파라미터명을 limit 에서 size 로 변경 #212

Byesol committed:
[BE] refactor: 상수 이름에 GROUPED 추가, 주석 위치 변경 #212

- GROUPED_CATEGORY_COUNT 로 변경
- HEARITS_PER_GROUPED_CATEGORY 로 변경
- 주석 위치 메서드 내부로 변경

Byesol committed:
[BE] refactor: repository의 파라미터 명 size로 변경 #212

yuyoungrhee reviewed (approved):
반영사항 확인했습니다! Approve 할게요

JO-eusan reviewed (approved):
변경 확인 했습니다. Approve할게요!
